sphinx/ext/autodoc/__init__.py

```diff
@@ -1056,6 +1057,61 @@ def add_directive_header(self, sig: str) -> None:
             self.add_line('   :async:', sourcename)
 
 
+class SingledispatchFunctionDocumenter(FunctionDocumenter):
+    """
+    Specialized Documenter subclass for singledispatch'ed functions.
+    """
+    objtype = 'singledispatch_function'
+    directivetype = 'function'
+    member_order = 30
+
+    # before FunctionDocumenter
+    priority = FunctionDocumenter.priority + 1
+
+    @classmethod
+    def can_document_member(cls, member: Any, membername: str, isattr: bool, parent: Any
+                            ) -> bool:
+        return (super().can_document_member(member, membername, isattr, parent) and
+                inspect.is_singledispatch_function(member))
+
+    def add_directive_header(self, sig: str) -> None:
+        sourcename = self.get_sourcename()
+
+        # intercept generated directive headers
+        with patch.object(self, 'add_line') as add_line:
+            super().add_directive_header(sig)
```

Comment:
I feel using mock is very hacky. But I don't have a good alternative for this. So I'm going with this until we'll find better solution.